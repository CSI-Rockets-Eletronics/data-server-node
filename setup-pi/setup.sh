#!/bin/bash -i

SETUP_FILES_DIR=$(realpath "$(dirname "$0")")
REPO_DIR=$(realpath "$(dirname "$0")/..")

# Exit if .env doesn't exist
if [ ! -f "$REPO_DIR/.env" ]; then
  echo "Please create a .env file in the root of the repository."
  echo "Set DATABASE_URL=\"postgresql://postgres@pass:5432/rockets-data?schema=public\""
  exit 1
fi

# Update the package list
sudo apt update

# Install PostgreSQL 15
sudo apt install -y postgresql-15

# Copy pg_hba.conf and postgresql.conf to the PostgreSQL config folder
cd "$SETUP_FILES_DIR"
sudo cp pg_hba.conf /etc/postgresql/15/main/
sudo cp postgresql.conf /etc/postgresql/15/main/

# Restart PostgreSQL to apply the new config
sudo systemctl restart postgresql

# Set PostgreSQL password for the postgres user
sudo -u postgres psql -c "ALTER USER postgres PASSWORD 'pass';"

# Install bun
curl -fsSL https://bun.sh/install | bash
source ~/.bashrc

# Install Node.js (necessary for Prisma)
sudo apt install -y nodejs

# Install dependencies and generate Prisma client
cd "$REPO_DIR"
bun install
bun prisma:generate

# Copy the service file to the systemd services folder
cd "$SETUP_FILES_DIR"
sudo cp data-server-node.service /etc/systemd/system/

# Reload systemd daemon to read the new service file
sudo systemctl daemon-reload

# Enable the service to run on boot
sudo systemctl enable data-server-node.service

# Start the service
sudo systemctl restart data-server-node.service
