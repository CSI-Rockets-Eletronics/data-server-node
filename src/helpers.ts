import process from 'node:process';
import {prisma} from './prisma';

export function ensureSystemClockIsSynchronizedOnLinux(): void {
	if (process.platform === 'linux') {
		try {
			const proc = Bun.spawnSync(['timedatectl', 'status']);
			const output = proc.stdout.toString();
			if (output.includes('System clock synchronized: yes')) {
				console.log('⏰ System clock is synchronized');
			} else {
				console.error(
					'⛔️ System clock is not synchronized! Run `timedatectl status` to see why.',
				);
			}
		} catch {
			console.warn(
				'⛔️ Unable to check if system clock is synchronized: timedatectl command not found',
			);
		}
	}
}

/**
 * If `sessionName` is undefined, defaults to the current session.
 */
export async function getSessionTimeRange(
	environmentKey: string,
	sessionName: string | undefined,
): Promise<{
	/** Inclusive */
	start: number | undefined;
	/** Inclusive */
	end: number | undefined;
}> {
	if (sessionName === undefined) {
		const curSession = await prisma.session.findFirst({
			where: {environmentKey},
			orderBy: {createdAt: 'desc'},
			select: {createdAt: true},
		});

		return {
			start: curSession ? Number(curSession.createdAt) : undefined,
			end: undefined,
		};
	}

	const session = await prisma.session.findUnique({
		where: {
			environmentKey_name: {
				environmentKey,
				name: sessionName,
			},
		},
		select: {createdAt: true},
	});

	if (!session) {
		throw new Error('Cannot find session');
	}

	const nextSession = await prisma.session.findFirst({
		where: {
			environmentKey,
			createdAt: {gt: session.createdAt},
		},
		orderBy: {createdAt: 'asc'},
		select: {createdAt: true},
	});

	return {
		start: Number(session.createdAt),
		end: nextSession ? Number(nextSession.createdAt) - 1 : undefined,
	};
}

export function curTimeMicros(): number {
	// On Linux, use the date command (to handle the case where the raspberry pi's
	// clock is not synchronized until after the server starts)
	if (process.platform === 'linux') {
		const proc = Bun.spawnSync(['date', '+%s%6N']);
		const output = proc.stdout.toString();
		return Number(output);
	}

	const milliseconds = performance.now() + performance.timeOrigin;
	return Math.round(milliseconds * 1000);
}
