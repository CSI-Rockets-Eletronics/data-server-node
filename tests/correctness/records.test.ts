import {describe, expect, test} from 'bun:test';
import {testNode} from '../test-node';
import {catchError} from '../helpers';
import {environmentKey} from '../setup';

describe('/records', () => {
	test('upload and get', async () => {
		const initial = await catchError(
			testNode.records.get({$query: {environmentKey, path: 'foo'}}),
		);
		expect(initial.records).toHaveLength(0);

		await catchError(
			testNode.records.post({
				environmentKey,
				path: 'foo',
				ts: 100,
				data: {bar100: 'baz100'},
			}),
		);

		expect(
			await catchError(
				testNode.records.get({$query: {environmentKey, path: 'foo'}}),
			),
		).toMatchSnapshot();

		await catchError(
			testNode.records.post({
				environmentKey,
				path: 'foo',
				ts: 200,
				data: {bar200: 'baz200'},
			}),
		);

		expect(
			await catchError(
				testNode.records.get({$query: {environmentKey, path: 'foo'}}),
			),
		).toMatchSnapshot();

		await catchError(
			testNode.records.post({
				environmentKey,
				path: 'foo',
				ts: 150,
				data: {bar150: 'baz150'},
			}),
		);

		expect(
			await catchError(
				testNode.records.get({$query: {environmentKey, path: 'foo'}}),
			),
		).toMatchSnapshot();

		// Shouldn't overwrite existing record with same ts
		await catchError(
			testNode.records.post({
				environmentKey,
				path: 'foo',
				ts: 100,
				data: {bar100: 'this should not appear'},
			}),
		);

		expect(
			await catchError(
				testNode.records.get({$query: {environmentKey, path: 'foo'}}),
			),
		).toMatchSnapshot();
	});

	test('upload batch', async () => {
		await catchError(
			testNode.records.batch.post({
				environmentKey,
				path: 'foo',
				records: [
					{ts: 100, data: {bar100: 'baz100'}},
					{ts: 200, data: {bar200: 'baz200'}},
					{ts: 150, data: {bar150: 'baz150'}},
				],
			}),
		);

		expect(
			await catchError(
				testNode.records.get({$query: {environmentKey, path: 'foo'}}),
			),
		).toMatchSnapshot();

		// Add some overlapping records, which should be ignored
		await catchError(
			testNode.records.batch.post({
				environmentKey,
				path: 'foo',
				records: [
					{ts: 200, data: {bar200: 'this should not appear'}},
					{ts: 150, data: {bar150: 'this should not appear'}},
					{ts: 300, data: {bar200: 'baz200'}},
					{ts: 300, data: {bar200: 'this should not appear'}},
				],
			}),
		);

		expect(
			await catchError(
				testNode.records.get({$query: {environmentKey, path: 'foo'}}),
			),
		).toMatchSnapshot();
	});

	test.each([false, true])(
		'upload and get from multiple sessions',
		async (batchUpload) => {
			const initialSessions = await catchError(
				testNode.sessions.get({$query: {environmentKey}}),
			);
			expect(initialSessions.sessions).toHaveLength(0);

			// Uploading a record should create a session
			await catchError(
				batchUpload
					? testNode.records.batch.post({
							environmentKey,
							path: 'foo',
							records: [{ts: 100, data: {bar100: 'baz100'}}],
					  })
					: testNode.records.post({
							environmentKey,
							path: 'foo',
							ts: 100,
							data: {bar100: 'baz100'},
					  }),
			);

			const sessionsAfterOne = await catchError(
				testNode.sessions.get({$query: {environmentKey}}),
			);
			expect(sessionsAfterOne.sessions).toHaveLength(1);

			// Manually create a new session
			await catchError(
				testNode.sessions.create.post({
					environmentKey,
				}),
			);

			const sessionsAfterTwo = await catchError(
				testNode.sessions.get({$query: {environmentKey}}),
			);
			expect(sessionsAfterTwo.sessions).toHaveLength(2);

			// Current session should have no records
			const curSessionRecords = await catchError(
				testNode.records.get({$query: {environmentKey, path: 'foo'}}),
			);
			expect(curSessionRecords.records).toHaveLength(0);

			const session1 = sessionsAfterTwo.sessions[0];
			const session2 = sessionsAfterTwo.sessions[1];

			// Also current session
			const session2Records = await catchError(
				testNode.records.get({
					$query: {environmentKey, path: 'foo', session: session2.session},
				}),
			);
			expect(session2Records.records).toHaveLength(0);

			// First session should have the record
			const session1Records = await catchError(
				testNode.records.get({
					$query: {environmentKey, path: 'foo', session: session1.session},
				}),
			);
			expect(session1Records.records).toHaveLength(1);

			// Add a record to the current session
			await catchError(
				batchUpload
					? testNode.records.batch.post({
							environmentKey,
							path: 'foo',
							records: [{ts: 200, data: {bar200: 'baz200'}}],
					  })
					: testNode.records.post({
							environmentKey,
							path: 'foo',
							ts: 200,
							data: {bar200: 'baz200'},
					  }),
			);

			// Current session should have the record
			const curSessionRecordsAfter = await catchError(
				testNode.records.get({
					$query: {environmentKey, path: 'foo'},
				}),
			);
			expect(curSessionRecordsAfter.records).toHaveLength(1);
			expect(curSessionRecordsAfter.records[0].ts).toBe(200);

			// Also current session
			const session2RecordsAfter = await catchError(
				testNode.records.get({
					$query: {environmentKey, path: 'foo', session: session2.session},
				}),
			);
			expect(session2RecordsAfter.records).toHaveLength(1);
			expect(session2RecordsAfter.records[0].ts).toBe(200);

			// First session should have only the first record
			const session1RecordsAfter = await catchError(
				testNode.records.get({
					$query: {environmentKey, path: 'foo', session: session1.session},
				}),
			);
			expect(session1RecordsAfter.records).toHaveLength(1);
			expect(session1RecordsAfter.records[0].ts).toBe(100);
		},
	);

	test('more complicated get', async () => {
		await testNode.records.batch.post({
			environmentKey,
			path: 'foo',
			records: [
				{ts: 100, data: {bar100: 'baz100'}},
				{ts: 200, data: {bar200: 'baz200'}},
			],
		});
		await testNode.records.batch.post({
			environmentKey,
			path: 'oof',
			records: [
				{ts: 100, data: {rab100: 'zab100'}},
				{ts: 200, data: {rab200: 'zab200'}},
			],
		});

		expect([
			await catchError(
				testNode.records.get({
					$query: {environmentKey, path: 'foo', startTs: '100'},
				}),
			),
			await catchError(
				testNode.records.get({
					$query: {environmentKey, path: 'foo', startTs: '100', take: '1'},
				}),
			),
			await catchError(
				testNode.records.get({
					$query: {environmentKey, path: 'foo', startTs: '200'},
				}),
			),
			await catchError(
				testNode.records.get({
					$query: {environmentKey, path: 'foo', startTs: '300'},
				}),
			),
		]).toMatchSnapshot();

		expect([
			await catchError(
				testNode.records.get({
					$query: {environmentKey, path: 'foo', endTs: '200'},
				}),
			),
			await catchError(
				testNode.records.get({
					$query: {environmentKey, path: 'foo', endTs: '200', take: '1'},
				}),
			),
			await catchError(
				testNode.records.get({
					$query: {environmentKey, path: 'foo', endTs: '100'},
				}),
			),
			await catchError(
				testNode.records.get({
					$query: {environmentKey, path: 'foo', endTs: '000'},
				}),
			),
		]).toMatchSnapshot();

		expect([
			await catchError(
				testNode.records.get({
					$query: {environmentKey, path: 'foo', startTs: '100', endTs: '200'},
				}),
			),
			await catchError(
				testNode.records.get({
					$query: {
						environmentKey,
						path: 'foo',
						startTs: '100',
						endTs: '200',
						take: '1',
					},
				}),
			),
			await catchError(
				testNode.records.get({
					$query: {environmentKey, path: 'foo', startTs: '200', endTs: '200'},
				}),
			),
			await catchError(
				testNode.records.get({
					$query: {environmentKey, path: 'foo', startTs: '100', endTs: '100'},
				}),
			),
			await catchError(
				testNode.records.get({
					$query: {environmentKey, path: 'foo', startTs: '200', endTs: '100'},
				}),
			),
		]).toMatchSnapshot();
	});
});
