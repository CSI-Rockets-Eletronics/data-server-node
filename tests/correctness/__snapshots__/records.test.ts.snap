// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`/records upload and get 1`] = `
{
  "records": [
    {
      "data": {
        "bar100": "baz100",
      },
      "ts": 100,
    },
  ],
}
`;

exports[`/records upload and get 2`] = `
{
  "records": [
    {
      "data": {
        "bar200": "baz200",
      },
      "ts": 200,
    },
    {
      "data": {
        "bar100": "baz100",
      },
      "ts": 100,
    },
  ],
}
`;

exports[`/records upload and get 3`] = `
{
  "records": [
    {
      "data": {
        "bar200": "baz200",
      },
      "ts": 200,
    },
    {
      "data": {
        "bar150": "baz150",
      },
      "ts": 150,
    },
    {
      "data": {
        "bar100": "baz100",
      },
      "ts": 100,
    },
  ],
}
`;

exports[`/records upload and get 4`] = `
{
  "records": [
    {
      "data": {
        "bar200": "baz200",
      },
      "ts": 200,
    },
    {
      "data": {
        "bar150": "baz150",
      },
      "ts": 150,
    },
    {
      "data": {
        "bar100": "baz100",
      },
      "ts": 100,
    },
  ],
}
`;

exports[`/records upload batch 1`] = `
{
  "records": [
    {
      "data": {
        "bar200": "baz200",
      },
      "ts": 200,
    },
    {
      "data": {
        "bar150": "baz150",
      },
      "ts": 150,
    },
    {
      "data": {
        "bar100": "baz100",
      },
      "ts": 100,
    },
  ],
}
`;

exports[`/records upload batch 2`] = `
{
  "records": [
    {
      "data": {
        "bar200": "baz200",
      },
      "ts": 300,
    },
    {
      "data": {
        "bar200": "baz200",
      },
      "ts": 200,
    },
    {
      "data": {
        "bar150": "baz150",
      },
      "ts": 150,
    },
    {
      "data": {
        "bar100": "baz100",
      },
      "ts": 100,
    },
  ],
}
`;

exports[`/records more complicated get 1`] = `
[
  {
    "records": [
      {
        "data": {
          "bar100": "baz100",
        },
        "ts": 100,
      },
      {
        "data": {
          "bar200": "baz200",
        },
        "ts": 200,
      },
    ],
  },
  {
    "records": [
      {
        "data": {
          "bar100": "baz100",
        },
        "ts": 100,
      },
    ],
  },
  {
    "records": [
      {
        "data": {
          "bar200": "baz200",
        },
        "ts": 200,
      },
    ],
  },
  {
    "records": [],
  },
]
`;

exports[`/records more complicated get 2`] = `
[
  {
    "records": [
      {
        "data": {
          "bar200": "baz200",
        },
        "ts": 200,
      },
      {
        "data": {
          "bar100": "baz100",
        },
        "ts": 100,
      },
    ],
  },
  {
    "records": [
      {
        "data": {
          "bar200": "baz200",
        },
        "ts": 200,
      },
    ],
  },
  {
    "records": [
      {
        "data": {
          "bar100": "baz100",
        },
        "ts": 100,
      },
    ],
  },
  {
    "records": [],
  },
]
`;

exports[`/records more complicated get 3`] = `
[
  {
    "records": [
      {
        "data": {
          "bar100": "baz100",
        },
        "ts": 100,
      },
      {
        "data": {
          "bar200": "baz200",
        },
        "ts": 200,
      },
    ],
  },
  {
    "records": [
      {
        "data": {
          "bar100": "baz100",
        },
        "ts": 100,
      },
    ],
  },
  {
    "records": [
      {
        "data": {
          "bar200": "baz200",
        },
        "ts": 200,
      },
    ],
  },
  {
    "records": [
      {
        "data": {
          "bar100": "baz100",
        },
        "ts": 100,
      },
    ],
  },
  {
    "records": [],
  },
]
`;
